File: day18/day18.py
#!/usr/bin/env python3

import re

class Node(object):
    PAIR = 0
    NUMBER = 1

    def __init__(self, parent, typ, number=None, left=None, right=None):
        self.parent = parent
        self.typ = typ
        self.number = number
        self.left = left
        if self.left:
            self.left.parent = self
        self.right = right
        if self.right:
            self.right.parent = self

    def __str__(self):
        if self.typ == self.PAIR:
            return "[{},{}]".format(self.left, self.right)
        else:
            return str(self.number)


PATTERN = re.compile(r"[\[\],]|\d+")
def parse(txt):
    lst = []
    for line in txt.splitlines():
        s = iter(PATTERN.findall(line))
        lst.append(parse_node(s, None))
    return lst

def parse_node(s, parent):
    value = next(s)
    if value == "[":
        n = Node(parent, Node.PAIR)
        n.left = parse_node(s, n)
        next(s)                 # ,
        n.right = parse_node(s, n)
        next(s)                 # ]
        return n
    else:
        return Node(parent, Node.NUMBER, number = int(value))

def add_left(n, val):
    while n.parent and n.parent.left == n:
        n = n.parent

    if not n.parent:
        return None

    n = n.parent.left
    while n and n.typ == n.PAIR:
        n = n.right

    if n:
        n.number += val

def add_right(n, val):
    while n.parent and n.parent.right == n:
        n = n.parent

    if not n.parent:
        return None

    n = n.parent.right
    while n and n.typ == n.PAIR:
        n = n.left

    if n:
        n.number += val

def explode(node, lvl):
    if lvl >= 4 and node.typ == node.PAIR:
        add_left(node, node.left.number)
        add_right(node, node.right.number)
        node.typ = node.NUMBER
        node.number = 0
        node.left = node.right = None
    elif node.typ == node.PAIR:
        explode(node.left, lvl+1)
        explode(node.right, lvl+1)
    else:
        pass

action = False
def split(node, lvl):
    global action
    if action:
        pass
    elif node.typ == node.PAIR:
        split(node.left, lvl+1)
        split(node.right, lvl+1)
    elif node.number >= 10:
        action = True
        node.typ = node.PAIR
        node.left = Node(node, node.NUMBER, node.number // 2)
        node.right = Node(node, node.NUMBER, (node.number + 1) // 2)
        explode(node, lvl)
    else:
        pass

def red(tree):
    global action
    explode(tree, 0)
    while True:
        action = False
        split(tree, 0)
        if not action:
            break

    return tree

def clone(a, parent):
    if a.typ == Node.NUMBER:
        return Node(parent, Node.NUMBER, number = a.number)
    else:
        n = Node(parent, Node.PAIR)
        n.left = clone(a.left, n)
        n.right = clone(a.right, n)
        return n

def add(a, b):
    n = Node(None, Node.PAIR)
    n.left = clone(a, n)
    n.right = clone(b, n)
    return red(n)

def magnitude(n):
    if n.typ == n.NUMBER:
        return n.number
    else:
        return 3 * magnitude(n.left) + 2 * magnitude(n.right)

def part1(numbers):
    a = numbers[0]
    for b in numbers[1:]:
        a = add(a, b)
    return magnitude(a)

def part2(numbers):
    maxsum = 0
    for i, a in enumerate(numbers):
        for j, b in enumerate(numbers):
            if i == j:
                continue

            m = magnitude(add(a, b))
            if maxsum < m:
                maxsum = m

    return maxsum

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        numbers = parse(txt)
        print("Part1:", part1(numbers))
        print("Part2:", part2(numbers))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

File: day18/day18.cpp
#include <algorithm>
#include <fstream>
#include <vector>
#include <memory>
#include <fmt/format.h>

using namespace std;

enum NodeType { PAIR, NUMBER };

struct Node
{
	NodeType type;
	int number;
	Node *up;
	unique_ptr<Node> left;
	unique_ptr<Node> right;

	Node(Node *parent, int number) :
		type(NUMBER),
		number(number),
		up(parent)
	{
	}

	explicit Node(Node *parent)
		: type(PAIR)
		, number(0)
		, up(parent)
	{
	}

	string to_string() const
	{
		if (type == PAIR)
		{
			return fmt::format(
				"[{},{}]",
				left->to_string(),
				right->to_string()
				);
		}
		else
		{
			return fmt::format("{}", number);
		}
	}
};

enum TokenType { LPAR, RPAR, COMMA, NUM, EOT, };

class Scanner
{
public:
	explicit Scanner(const string& str) :
		cur(str.cbegin()),
		end(str.cend()),
		type(EOT),
		value(0)
	{
		get();
	}

	void get()
	{
		value = 0;
		if (cur == end)
		{
			type = EOT;
			return;
		}
		switch (*cur)
		{
		case '[': ++cur; type = LPAR; break;
		case ']': ++cur; type = RPAR; break;
		case ',': ++cur; type = COMMA; break;
		default:
			if (isdigit(*cur))
			{
				while (cur != end && isdigit(*cur))
				{
					value = value * 10 + *cur - '0';
					++cur;
				}
				type = NUM;
			}
			else
			{
				cur = end;
				type = EOT;
			}
		}
	}

	void match(TokenType t)
	{
		if (type != t)
		{
			throw std::runtime_error(
				fmt::format("expected {}, found {}", t, type));
		}
		get();
	}

	TokenType getToken() const { return type; }
	int getValue() const { return value; }

private:
	string::const_iterator cur;
	string::const_iterator end;
	TokenType type;
	int value;
};

static unique_ptr<Node> parse_node(Scanner &s, Node *parent)
{
	if (s.getToken() == LPAR)
	{
		auto n = make_unique<Node>(parent);
		s.get();
		n->left = parse_node(s, n.get());
		s.match(COMMA);
		n->right = parse_node(s, n.get());
		s.match(RPAR);
		return n;
	}
	else
	{
		auto n = make_unique<Node>(parent, s.getValue());
		s.get();
		return n;
	}
}

static unique_ptr<Node> parse(const string& str)
{
	Scanner s(str);
	return parse_node(s, nullptr);
}

static void add_left(Node *n, int value)
{
	while (n->up && n->up->left.get() == n)
	{
		n = n->up;
	}

	if (n->up)
	{
		n = n->up->left.get();
		while (n && n->type == PAIR)
		{
			n = n->right.get();
		}
		if (n)
		{
			n->number += value;
		}
	}
}

static void add_right(Node *n, int value)
{
	while (n->up && n->up->right.get() == n)
	{
		n = n->up;
	}

	if (n->up)
	{
		n = n->up->right.get();
		while (n && n->type == PAIR)
		{
			n = n->left.get();
		}
		if (n)
		{
			n->number += value;
		}
	}
}

static void explode(Node& n, int lvl)
{
	if (lvl >= 4 && n.type == PAIR)
	{
		add_left(&n, n.left->number);
		add_right(&n, n.right->number);

		n.left.reset();
		n.right.reset();
		n.type = NUMBER;
		n.number = 0;
	}
	else if (n.type == PAIR)
	{
		explode(*n.left, lvl + 1);
		explode(*n.right, lvl + 1);
	}
	else
	{
	}
}

bool action;
static void split(Node& n, int lvl)
{
	if (action)
	{
		return;
	}
	else if (n.type == PAIR)
	{
		split(*n.left, lvl+1);
		split(*n.right, lvl+1);
	}
	else if (n.number >= 10)
	{
		n.type = PAIR;
		n.left = make_unique<Node>(&n, n.number / 2);
		n.right = make_unique<Node>(&n, (n.number + 1) / 2);
		explode(n, lvl);
		action = true;
	}
}

static void reduce(Node& n)
{
	explode(n, 0);
	do
	{
		action = false;
		split(n, 0);
	} while (action);
}

static unique_ptr<Node> clone(const Node& n, Node *parent)
{
	if (n.type == NUMBER)
	{
		return make_unique<Node>(parent, n.number);
	}
	else
	{
		auto c = make_unique<Node>(parent);
		c->left = clone(*n.left, c.get());
		c->right = clone(*n.right, c.get());
		return c;
	}
}

static unique_ptr<Node> add(const Node &a, const Node &b)
{
	auto n = make_unique<Node>(nullptr);
	n->left = clone(a, n.get());
	n->right = clone(b, n.get());
	reduce(*n);
	return n;
}

static int magnitude(const Node &n)
{
	if (n.type == NUMBER)
	{
		return n.number;
	}
	else
	{
		return 3 * magnitude(*n.left)
			+ 2 * magnitude(*n.right);
	}
}

static int part1(const vector<unique_ptr<Node>>& numbers)
{
	auto acc = clone(*numbers.at(0), nullptr);
	for (auto it = numbers.begin() + 1;
	     it != numbers.end();
	     ++it)
	{
		acc = add(*acc, **it);
	}
	return magnitude(*acc);
}

static int part2(const vector<unique_ptr<Node>>& numbers)
{
	int max = 0;
	for (auto i = numbers.begin();
	     i != numbers.end();
	     ++i)
	{
		for (auto j = numbers.begin();
		     j != numbers.end();
		     ++j)
		{
			if (i != j)
			{
				auto number = add(**i, **j);
				auto m = magnitude(*number);
				if (max < m)
				{
					max = m;
				}
			}
		}
	}
	return max;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<unique_ptr<Node>> numbers;
	string line;
	while (getline(in, line))
	{
		numbers.push_back(parse(line));
	}
	in.close();

	fmt::print("Part1: {}\n", part1(numbers));
	fmt::print("Part2: {}\n", part2(numbers));
	return 0;
}
